src/dmd/mars.d

```diff
@@ -2125,3 +2168,278 @@ private bool parseCommandLine(const ref Strings arguments, const size_t argc, re
     return errors;
 }
 
+
+private __gshared bool includeImports = false;
+// array of module patterns used to include/exclude imported modules
+private __gshared Array!(const(char)*) includeModulePatterns;
+private __gshared Modules compiledImports;
+private extern(C++) bool marsOnImport(Module m)
+{
+    if (includeImports)
+    {
+        Identifiers empty;
+        if (includeImportedModuleCheck(ModuleComponentRange(
+            (m.md && m.md.packages) ? m.md.packages : &empty, m.ident, m.isPackageFile)))
+        {
+            if (global.params.verbose)
+                fprintf(global.stdmsg, "compileimport (%s)\n", m.srcfile.toChars);
+            compiledImports.push(m);
+            return true; // this import will be compiled
+        }
+    }
+    return false; // this import will not be compiled
+}
+
+// A range of component identifiers for a module
+private struct ModuleComponentRange
+{
+    Identifiers* packages;
+    Identifier name;
+    bool isPackageFile;
+    size_t index;
+    @property auto totalLength() const { return packages.dim + 1 + (isPackageFile ? 1 : 0); }
+
+    @property auto empty() { return index >= totalLength(); }
+    @property auto front() const
+    {
+        if (index < packages.dim)
+            return (*packages)[index];
+        if (index == packages.dim)
+            return name;
+        else
+            return Identifier.idPool("package");
```

Comment:
This should probably just return `Id.package`, I'd be surprised if `package` wasn't already a pre-allocated identifier.